src/main/java/com/example/bookstorewebapp/repository/book/BookSpecificationBuilder.java

```diff
@@ -0,0 +1,33 @@
+package com.example.bookstorewebapp.repository.book;
+
+import com.example.bookstorewebapp.dto.BookSearchParameters;
+import com.example.bookstorewebapp.model.Book;
+import com.example.bookstorewebapp.repository.SpecificationBuilder;
+import com.example.bookstorewebapp.repository.SpecificationProviderManager;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    @Autowired
+    private SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> defaultSpec = Specification.where(null);
+        if (searchParameters.titles() != null && searchParameters.titles().length > 0) {
+            defaultSpec = defaultSpec
+                    .and(bookSpecificationProviderManager.getSpecificationProvider("title")
+                            .getSpecification(searchParameters.titles()));
+        }
+        if (searchParameters.authors() != null && searchParameters.authors().length > 0) {
+            defaultSpec = defaultSpec
+                    .and(bookSpecificationProviderManager.getSpecificationProvider("author")
+                            .getSpecification(searchParameters.authors()));
+        }
+        return defaultSpec;
+    }
```

Comment:
I see at least 2 code duplications:

1. Check if the collection is null or empty. I suggest to create some util class for such checks or use already existing libraries like Apache common utils
2. Code that populates specification is common for both author and title, so I suggest extracting it in method and reuse. This will make code more flexible because it will be much easier to add new search parameters, maybe event strategy will be suitable here but it may be overcomplication, so just make a method for now